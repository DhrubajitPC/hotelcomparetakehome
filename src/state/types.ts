export type Hotel = {
  id: number;
  name: string;
  rating: number;
  stars: number;
  address: string;
  photo: string;
  description: string;
};

export type Price = {
  id: number;
  price: number;
  competitors?: Record<string, number>;
  taxes_and_fees?: {
    tax: number;
    hotel_fees: number;
  };
};

export type Currency = "USD" | "SGD" | "CNY" | "KRW";

type LoadingState = "REQUEST" | "FAILURE" | "SUCCESS";

export type State = {
  hotels: Array<Hotel>;
  prices: Array<Price>;
  selectedCurrency: Currency;
  loading: {
    hotels: LoadingState;
    prices: LoadingState;
  };
};
