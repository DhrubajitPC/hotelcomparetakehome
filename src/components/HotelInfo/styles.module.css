
.container {
	display: flex;
	justify-content: left;
	border: 1px solid var(--black);
	border-radius: 10px;
	margin: 10px 0;
	padding: 20px;
	color: var(--black)
}

.imageContainer {
	display: flex;
	align-items: center;
	margin-right: 30px;
}
.imageContainer > img{
	width: 300px;
	height: 300px;
}

.body {
	margin-right: auto;
	padding: 0 50px;
}

.review {
	min-width: 100px;
}