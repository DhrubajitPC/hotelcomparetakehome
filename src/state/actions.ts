export const REPLACE = "REPLACE";
export const FETCH_PRICE_REQUEST = "REPLACE";
export const FETCH_PRICE_SUCCESS = "FETCH_PRICE_SUCCESS";
export const FETCH_PRICE_FAILURE = "FETCH_PRICE_FAILURE";
