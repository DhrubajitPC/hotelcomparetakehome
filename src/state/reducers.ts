import { Reducer } from "react";
import * as A from "./actions";
import { State } from "./types";

export const reducer: Reducer<State, { type: string; payload: any }> = (
  draft,
  action
) => {
  switch (action.type) {
    case A.REPLACE:
      return action.payload;
  }
};
