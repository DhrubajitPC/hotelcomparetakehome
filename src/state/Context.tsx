import React, { FC, useContext, useEffect } from "react";
import { useImmerReducer } from "use-immer";
import { apiWrapper } from "../infra/api";
import { reducer } from "./reducers";
import { State } from "./types";

type Context = {
  state: State;
  dispatch: React.Dispatch<{
    type: string;
    payload: any;
  }>;
};

const initialState: Context = {
  state: {
    hotels: [],
    prices: { USD: [], CNY: [], KRW: [], SGD: [] },
    loading: { hotels: "REQUEST", prices: "REQUEST" },
    selectedCurrency: "USD",
  },
  dispatch: () => {},
};

const AppContext = React.createContext<Context>(initialState);

export const useAppContext = () => useContext(AppContext);

export const useAPI = () => {
  const {dispatch} = useContext(AppContext)
  return apiWrapper(dispatch)
}

const AppContextProvider: FC = ({ children }) => {
  const [state, dispatch] = useImmerReducer(reducer, initialState.state);
  const api = apiWrapper(dispatch)

  // fetch initial list of hotels and usd currency
  useEffect(() => {
    function fetchInitData() {
      api.fetchHotels()
      api.fetchPrices('USD')
    }
    fetchInitData();
  }, []);

  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export default AppContextProvider;
