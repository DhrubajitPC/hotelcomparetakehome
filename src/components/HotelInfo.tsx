import React, { FC } from "react";
import { Hotel } from "../state/types";

export type Props = {
  onBook: () => void;
  price: number;
  currency: string;
} & Omit<Hotel, "id">;

const HotelInfo: FC<Props> = ({
  address,
  description,
  name,
  photo,
  rating,
  stars,
  onBook,
  price,
  currency,
  children,
}) => (
  <>
    <div
      style={{
        display: "flex",
        justifyContent: "left",
      }}
    >
      <div>
        <img alt="hotel" src={photo} />
      </div>
      <div
        style={{
          marginRight: "auto",
        }}
      >
        <h3>{name}</h3>
        <p>{address}</p>
        <div dangerouslySetInnerHTML={{ __html: description }} />
        <button onClick={onBook}>Book!</button>
      </div>
      <div>
        <p>{stars} star Hotel</p>
        <p>Rating {rating}</p>
        <p>
          {currency} {price}
        </p>
      </div>
    </div>
    {children && <div>{children}</div>}
  </>
);

export default HotelInfo;
