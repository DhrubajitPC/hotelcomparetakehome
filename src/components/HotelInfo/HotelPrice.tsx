import { FC } from "react";
import { useAppContext } from "../../state/Context";
import TaxBreakdown from "../TaxBreakdown";
import Tooltip from "../Tooltip";
import { usePriceBreakdown } from "./hooks";
import { formatPrice } from "./utils";

const HotelPrice: FC<{ hotelId: number }> = ({ hotelId }) => {
  const { state } = useAppContext();
  const currency = state.selectedCurrency;
  const priceBreakdown = usePriceBreakdown(hotelId);

  const isTaxInclusive = Boolean(priceBreakdown.hotelFees);
  const showCurrency = priceBreakdown.totalPrice !== "Rates Unavailable";

  let price;
  if (showCurrency) {
    price = formatPrice(currency, priceBreakdown.totalPrice as number);
  } else {
    price = priceBreakdown.totalPrice;
  }

  return isTaxInclusive ? (
    <Tooltip
      body={
        <TaxBreakdown
          tax={priceBreakdown.tax!}
          hotelFees={priceBreakdown.hotelFees!}
        />
      }
    >
      <p style={{ textDecoration: "underline" }}>
        {showCurrency && currency} {price}*
      </p>
    </Tooltip>
  ) : (
    <p>
      {showCurrency && currency} {price}
    </p>
  );
};

export default HotelPrice;
