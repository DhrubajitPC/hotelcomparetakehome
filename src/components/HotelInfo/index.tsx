import React, { FC } from "react";
import { useComparisons } from "../../hooks";
import { Hotel } from "../../state/types";
import ComparePrices from "../ComparePrices";
import HotelPrice from "../HotelPrice";
import styles from "./styles.module.css";

const HotelInfo: FC<Hotel> = ({
  id,
  address,
  description,
  name,
  photo,
  rating,
  stars,
}) => {
  const showHr = useComparisons(id).length > 0;
  return (
    <>
      <div className={styles.container}>
        <div className={styles.imageContainer}>
          <img alt="hotel" src={photo} />
        </div>
        <div className={styles.body}>
          <h3>{name}</h3>
          <p>
            <em>Address:</em> {address}
          </p>
          <div dangerouslySetInnerHTML={{ __html: description }} />
          {showHr && <hr />}
          <ComparePrices hotelId={id} />
        </div>
        <div className={styles.review}>
          <h4>Review</h4>
          <p>{stars} ‚≠ê Hotel</p>
          <p>Rating: {rating}</p>
          <HotelPrice hotelId={id} />
        </div>
      </div>
      {/* {children && <div>{children}</div>} */}
    </>
  );
};

export default HotelInfo;
