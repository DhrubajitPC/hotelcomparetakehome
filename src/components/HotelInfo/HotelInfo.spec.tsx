import { render } from "@testing-library/react";
import React from "react";
import HotelInfo from ".";
import { Hotel } from "../../state/types";

describe("#HotelInfo", () => {
  let props: Hotel;
  beforeAll(() => {
    props = {
      id: 1,
      address: "some address",
      name: "Shinagawa Prince Hotel",
      rating: 7.7,
      stars: 4,
      photo: "photo url",
      description: "some mock description",
    };
  });
  it("should render correctly", () => {
    const { container } = render(<HotelInfo {...props} />);
    expect(container).toMatchSnapshot();
  });
});
