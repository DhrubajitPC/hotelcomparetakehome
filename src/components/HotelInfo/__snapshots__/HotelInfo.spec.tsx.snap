// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#HotelInfo should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="imageContainer"
    >
      <img
        alt="hotel"
        src="photo url"
      />
    </div>
    <div
      class="body"
    >
      <h3>
        Shinagawa Prince Hotel
      </h3>
      <p>
        <em>
          Address:
        </em>
         
        some address
      </p>
      <div>
        some mock description
      </div>
    </div>
    <div
      class="review"
    >
      <h4>
        Review
      </h4>
      <p>
        4
         ‚≠ê Hotel
      </p>
      <p>
        Rating: 
        7.7
      </p>
      <p
        style="color: rgb(235, 77, 75);"
      >
        Price: 
         
        Rates Unavailable
      </p>
      <button
        style="padding: 20px; background: rgb(72, 52, 212); border-radius: 10px; font-size: 18px; color: rgb(255, 255, 255); cursor: pointer;"
      >
        Book!
      </button>
    </div>
  </div>
</div>
`;
